### ISSUE TITLE
unhelpful diagnostic for attempted static member dereference in constructor call

### ISSUE THREAD

**Issue by ojhunt** [llvm/llvm-project#222428]
I haven't worked out the exact thing that is going wrong so the title is just a description of the symptom :D

```cpp
struct S { S(int); };
struct A { static int *c; };
void f() { S(*A::c); }
```

There's disagreement as to whether this is accepted (EDG rejects, gcc accepts).

Both clang and msvc complain about `c` being redeclared. O_o

**Comment by zygoloid** [comment]
That is not a dereference, it's a declaration of a variable whose type is `S *` and whose name is `A::c`, just like
```c++
S *A::c;
```
The fact that you can't declare a qualified name at block scope is not considered by the disambiguation rule, so this disambiguates as a variable declaration, not as an expression.

Clang says:
```console
<source>:3:18: error: definition or redeclaration of 'c' not allowed inside a function
    3 | void f() { S(*A::c); }
      |               ~~~^
```
... which is technically correct -- you can't redeclare `A::c` there -- but unhelpful. We should detect that the parse as a declaration was ambiguous with a parse as an expression (IIRC we already track that in a bit on the `DeclSpec`) and produce a better diagnostic. In this case, you can disambiguate by adding parentheses around the whole expression:
```c++
(S(*A::c));
```
... or by casting to void:
```c++
static_cast<void>(S(*A::c));
```

**Comment by ojhunt** [comment]
O_o

That is a new syntax I was not previously aware of.

O_o.
